orttraining/orttraining/training_ops/cuda/communication/send.cc

```diff
@@ -3,69 +3,117 @@
-      Send<T>);
-
-REGISTER_SEND_KERNEL_TYPED(MLFloat16)
-REGISTER_SEND_KERNEL_TYPED(float)
-REGISTER_SEND_KERNEL_TYPED(double)
+ONNX_OPERATOR_KERNEL_EX(
+    Send,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeCPUInput>(0)   /* CPU variable */
+        .InputMemoryType<OrtMemTypeCPUInput>(1)   /* CPU variable */
+        .OutputMemoryType<OrtMemTypeCPUOutput>(0) /* CPU variable */
+        .TypeConstraint("TInt64", DataTypeImpl::GetTensorType<int64_t>())
+        .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>())
+        .TypeConstraint("V", DataTypeImpl::AllFixedSizeTensorTypes()),
+    Send);
 
 void CUDART_CB HostSend(void* args) {
   CommInfo_t* info = reinterpret_cast<CommInfo_t*>(args);
-  int mpi_code  = MPI_Send(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD);
+  int mpi_code = MPI_Send(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD);
   ORT_ENFORCE(mpi_code == MPI_SUCCESS, "MPI Send fails.");
 }
 
-template <typename T>
-Status Send<T>::ComputeInternal(OpKernelContext* ctx) const {
+Status Send::ComputeInternal(OpKernelContext* ctx) const {
+  // Extract Remote rank
+  const Tensor* remote_rank_tensor = ctx->Input<Tensor>(0);
+  const int64_t* remote_rank = remote_rank_tensor->template Data<int64_t>();
+  int dst = static_cast<int>(*remote_rank);
+
   // Check if control signal is true.
-  const Tensor* input_signal_tensor = ctx->Input<Tensor>(0);
+  const Tensor* input_signal_tensor = ctx->Input<Tensor>(1);
   const bool* input_signal = input_signal_tensor->template Data<bool>();
   ORT_ENFORCE(*input_signal, "Input control signal of Send must be true before executing the node.");
 
-  // Start the communication.
-  const Tensor* x_tensor = ctx->Input<Tensor>(1);
-  size_t tensor_shape_size = static_cast<size_t>(x_tensor->Shape().NumDimensions());;
-  size_t tensor_size = static_cast<size_t>(x_tensor->Shape().Size());
-
+  // Create buffers
+  int tensor_num = static_cast<int>(element_types_.size());
+  // TODO move the following variables to member variables for extending life-time
+  // if we want to make the entire call async
+  std::vector<size_t> prefix_tensor_shape_sizes;
+  std::vector<int64_t> aggregated_tensor_shapes;
+  size_t aggregated_aligned_tensor_bytes = 0;
+  size_t alignment = 256;
+  std::vector<size_t> tensor_offsets_in_bytes;
+  std::vector<size_t> tensor_sizes_in_bytes;
+
+  // Compute tensor shapes and sizes
+  size_t sum = 0;
+  for (int i = 0; i < tensor_num; ++i) {
+    const Tensor* x_tensor = ctx->Input<Tensor>(i + 2);
+    sum += x_tensor->Shape().NumDimensions();
+    prefix_tensor_shape_sizes.push_back(sum);
+    aggregated_tensor_shapes.insert(aggregated_tensor_shapes.end(),
+                                    x_tensor->Shape().GetDims().begin(),
+                                    x_tensor->Shape().GetDims().end());
+
+    // handle alignment requirement
```

Comment:
What kind of requirement? Is it for cudaMemcopy?

Reply:
Cuda memory object has alignment, and the perf of cudamemcpy might be different.
In most devices, pinned memory,  it is 256 bytes.
I think ORT also use it somewhere.
I just give this option. 
We changing it to 1 first, we want to save memory.